src/EditorFeatures/Core/Workspaces/AbstractTextBufferVisibilityTracker.cs

```diff
@@ -0,0 +1,227 @@
+            /// </summary>
+            public ImmutableHashSet<Action> Callbacks { get; private set; } = ImmutableHashSet<Action>.Empty;
+
+            public VisibleTrackerData(
+                AbstractTextBufferVisibilityTracker<TTextView, TVisibilityChangedCallback> tracker,
+                ITextBuffer subjectBuffer)
+            {
+                _tracker = tracker;
+                _subjectBuffer = subjectBuffer;
+                _visibilityChangedCallback = tracker.GetVisiblityChangeCallback(this);
+
+                UpdateAssociatedViews();
+            }
+
+            public void Dispose()
+            {
+                _tracker._threadingContext.ThrowIfNotOnUIThread();
+
+                // Shouldn't be disposing of this if we still have clients that want to hear about visibility changes.
+                Contract.ThrowIfTrue(Callbacks.Count > 0);
+
+                // Clear out all our textviews.  This will disconnect us from any events we have registered with them.
+                UpdateTextViews(Array.Empty<ITextView>());
+
+                Contract.ThrowIfTrue(TextViews.Count > 0);
+            }
+
+            public void AddCallback(Action callback)
+            {
+                _tracker._threadingContext.ThrowIfNotOnUIThread();
+                this.Callbacks = this.Callbacks.Add(callback);
+            }
+
+            public void RemoveCallback(Action callback)
+            {
+                _tracker._threadingContext.ThrowIfNotOnUIThread();
+                this.Callbacks = this.Callbacks.Remove(callback);
+            }
+
+            public void UpdateAssociatedViews()
+            {
+                _tracker._threadingContext.ThrowIfNotOnUIThread();
+
+                // Update us to whatever the currently associated text views are for this buffer.
+                UpdateTextViews(_tracker._associatedViewService.GetAssociatedTextViews(_subjectBuffer));
+            }
+
+            private void UpdateTextViews(IEnumerable<ITextView> associatedTextViews)
+            {
+                var removedViews = TextViews.Except(associatedTextViews);
+                var addedViews = associatedTextViews.Except(TextViews);
+
+                // Disconnect from hearing about visibility changes for any views we're no longer associated with.
+                foreach (var removedView in removedViews)
+                {
+                    if (removedView is TTextView genericView)
+                        _tracker.RemoveVisibilityChangedCallback(genericView, _visibilityChangedCallback);
+                }
+
+                // Connect to hearing about visbility changes for any views we are associated with.
+                foreach (var addedView in addedViews)
+                {
+                    if (addedView is TTextView genericView)
+                        _tracker.AddVisibilityChangedCallback(genericView, _visibilityChangedCallback);
```

Comment:
defer to platform specific code to attach/detach the visibility changed listener.